tsdb/tsi1/partition.go

```diff
@@ -1294,40 +1291,75 @@ func (p *Partition) compactLogFile(ctx context.Context, logFile *LogFile, interr
 }
 
 // MeasurementCardinalityStats returns cardinality stats for all measurements.
-func (p *Partition) MeasurementCardinalityStats() MeasurementCardinalityStats {
+func (p *Partition) MeasurementCardinalityStats() (MeasurementCardinalityStats, error) {
 	p.mu.RLock()
 	defer p.mu.RUnlock()
 
-	stats := p.stats.Clone()
+	// Return cached version, if enabled and the TTL is less than the last cache time.
+	if p.StatsTTL > 0 && !p.lastStatsTime.IsZero() && time.Since(p.lastStatsTime) < p.StatsTTL {
+		return p.statsCache.Clone(), nil
+	}
 
-	if p.activeLogFile != nil {
-		stats.Add(p.activeLogFile.MeasurementCardinalityStats())
+	// If cache is unavailable then generate fresh stats.
+	stats, err := p.measurementCardinalityStats()
+	if err != nil {
+		return nil, err
 	}
-	return stats
-}
 
-// ComputeMeasurementCardinalityStats computes cardinality stats from raw data.
-func (p *Partition) ComputeMeasurementCardinalityStats() (MeasurementCardinalityStats, error) {
-	p.mu.RLock()
-	defer p.mu.RUnlock()
+	// Cache the stats if enabled.
+	if p.StatsTTL > 0 {
+		p.statsCache = stats
+		p.lastStatsTime = time.Now()
+	}
 
+	return stats, nil
+}
+
+func (p *Partition) measurementCardinalityStats() (MeasurementCardinalityStats, error) {
 	fs, err := p.fileSet.Duplicate()
 	if err != nil {
 		return nil, err
 	}
 	defer fs.Release()
-	return fs.Stats(), nil
-}
 
-// computeStats calculates the measurement stats from all files except the active log.
-// FileSet must already be retained when calling this function.
-func (p *Partition) computeStats() {
-	// Shallow copy the fileset and trim initial active log file.
-	fs := *p.fileSet
-	fs.files = fs.files[1:]
+	stats := make(MeasurementCardinalityStats)
+	mitr := fs.MeasurementIterator()
+	if mitr == nil {
+		return stats, nil
+	}
+
+	for {
+		// Iterate over each measurement and set cardinality.
+		mm := mitr.Next()
+		if mm == nil {
+			return stats, nil
+		}
+
+		// Obtain all series for measurement.
+		sitr := fs.MeasurementSeriesIDIterator(mm.Name())
+		if sitr == nil {
+			continue
+		}
+
+		// All iterators should be series id set iterators except legacy 1.x data.
```

Comment:
Will this be a concern when users eventually upgrade from 1.x to 2.x? Want to make sure silently skipping is appropriate.

Reply:
This has been in there for a while. IIRC, we decided that users would either need to wait for compactions to update their files automatically or do a manual rebuild if they want to see org/bucket cardinality metrics. Although I'm not sure we even expose those in OSS right now.